In [1]:
#importing basic python Libraries for data science work 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
sc_2017 = pd.ExcelFile('2017.xlsx')
months = np.array(sc_2017.sheet_names)
# Create a 2017 dataframe to append all sheets into
sc_17 = pd.DataFrame()


for sheet in months:
    
    # Read in the dataframe... 
    df = pd.read_excel(sc_2017, sheet)
    
    # Append the dataframe onto our 2017 dataframe
    sc_17 = pd.concat([sc_17, df])

In [3]:
sc_17

Report Date  Route      Time     Day                 Location  \
0     2017-01-01    509  00:49:00  Sunday  Spadina and Queens Quay   
1     2017-01-01    501  01:07:00  Sunday    Lakeshore and Kipling   
2     2017-01-01    505  01:09:00  Sunday   Dundas and Howard Park   
3     2017-01-01    506  01:15:00  Sunday                High Park   
4     2017-01-01    506  01:18:00  Sunday               Parliament   
...          ...    ...       ...     ...                      ...   
1693  2017-12-31    501  19:58:00  Sunday                    Queen   
1694  2017-12-31    501  21:15:00  Sunday   Queen and Neville park   
1695  2017-12-31    501  21:44:00  Sunday       Queen and Dufferin   
1696  2017-12-31    510  22:15:00  Sunday         Spadina and King   
1697  2017-12-31    505  22:30:00  Sunday    Broadview and Gerrard   

                Incident  Min Delay  Min Gap Direction  Vehicle  
0             Mechanical        7.0     14.0       N/B   4426.0  
1             Mechanical       32.0     44.0       W/B   4071.0  
2             Mechanical       10.0     20.0       W/B   4179.0  
3             Mechanical       10.0     20.0       E/B   4059.0  
4     Emergency Services       10.0     20.0       E/B   4106.0  
...                  ...        ...      ...       ...      ...  
1693       General Delay        7.0     15.0       B/W      NaN  
1694          Mechanical       12.0     24.0       W/B   4110.0  
1695          Mechanical       12.0     24.0       W/B   1419.0  
1696          Mechanical        5.0     10.0       S/B   4447.0  
1697          Mechanical        8.0     16.0       N/B   4167.0  

[13762 rows x 10 columns]

In [4]:
sc_2018 = pd.ExcelFile('2018.xlsx')
months = np.array(sc_2018.sheet_names)
# Create a 2018 dataframe to append all sheets into
sc_18 = pd.DataFrame()


for sheet in months:
    
    # Read in the dataframe 
    df = pd.read_excel(sc_2018, sheet)
    
    # Append the dataframe onto our 2018 dataframe
    sc_18 = pd.concat([sc_18, df])

In [5]:
sc_18

Report Date  Route      Time     Day                   Location  \
0    2018-01-01    512  00:11:00  Monday           St Clair station   
1    2018-01-01    501  00:14:00  Monday     Queen and Neville loop   
2    2018-01-01    501  01:51:00  Monday  Lakeshore and Long branch   
3    2018-01-01    301  03:14:00  Monday            Queen and Roncy   
4    2018-01-01    505  03:27:00  Monday        Dundas West Station   
..          ...    ...       ...     ...                        ...   
909  2018-12-31    504  21:53:00  Monday                   Berkeley   
910  2018-12-31    506  22:05:00  Monday   Coxwell and Upper Gerard   
911  2018-12-31    501  22:09:00  Monday               Neville Loop   
912  2018-12-31    501  23:23:00  Monday          Spadina/Lakeshore   
913  2018-12-31    506  23:33:00  Monday               Main Station   

          Incident  Min Delay  Min Gap Direction  Vehicle  
0       Mechanical       10.0     20.0       W/B   4441.0  
1       Mechanical       12.0     24.0       E/B   4091.0  
2       Mechanical        8.0     16.0       W/B   8039.0  
3       Mechanical       12.0     24.0       W/B   4024.0  
4       Mechanical       42.0     53.0       E/B   4004.0  
..             ...        ...      ...       ...      ...  
909        Held By       18.0     23.0       W/B   4464.0  
910     Mechanical        8.0     14.0       E/B   4150.0  
911     Mechanical        8.0     17.0       E/B   4024.0  
912  Investigation        8.0     17.0       S/B   4066.0  
913     Mechanical        8.0     16.0       W/B   4176.0  

[15612 rows x 10 columns]

In [6]:
sc_2019 = pd.ExcelFile('2019.xlsx')
months = np.array(sc_2019.sheet_names)
# Create a 2019 dataframe to append all sheets into
sc_19 = pd.DataFrame()


for sheet in months:
    
    # Read in the dataframe 
    df = pd.read_excel(sc_2019, sheet)
    
    # Append the dataframe onto our 2019 dataframe
    sc_19 = pd.concat([sc_19, df])

In [7]:
sc_19

Report Date  Route      Time      Day                Location  \
0    2019-01-01    301  01:08:00  Tuesday         Queen/Braodview   
1    2019-01-01    511  01:30:00  Tuesday        Bathurst/College   
2    2019-01-01    306  01:40:00  Tuesday        Dundas West stn.   
3    2019-01-01    505  03:10:00  Tuesday    Lansdowne and Dundas   
4    2019-01-01    310  03:55:00  Tuesday    Spadina and Lakshore   
..          ...    ...       ...      ...                     ...   
810  2019-12-31    505  20:00:00  Tuesday              Parliament   
811  2019-12-31    505  20:32:00  Tuesday       Broadview Station   
812  2019-12-31    501  22:32:00  Tuesday             Humber Loop   
813  2019-12-31    506  23:20:00  Tuesday  Carlton and Parliament   
814  2019-12-31    511  23:36:00  Tuesday        Bathurst Station   

               Incident  Min Delay  Min Gap Direction  Vehicle  Incident ID  
0               Held By        6.0     13.0       E/B   4193.0          NaN  
1         Investigation        5.0     10.0       N/B   1038.0          NaN  
2            Mechanical        8.0     16.0       W/B   4146.0          NaN  
3            Mechanical        6.0     12.0       E/B   8416.0          NaN  
4               Held By       20.0     30.0       N/B   4465.0          NaN  
..                  ...        ...      ...       ...      ...          ...  
810       General Delay       18.0     23.0       W/B   1262.0          NaN  
811  Emergency Services        5.0     10.0       W/B   3104.0          NaN  
812             Held By        5.0     10.0       E/B   4576.0          NaN  
813       General Delay       20.0     26.0       E/B   4498.0          NaN  
814             Held By        7.0     14.0       S/B   4583.0          NaN  

[11882 rows x 11 columns]

In [8]:
#Reset Index
sc_17.reset_index(inplace = True, drop = True )
sc_18.reset_index(inplace = True, drop = True )
sc_19.reset_index(inplace = True, drop = True )

In [9]:
#saving all df in an array
a = [sc_17,sc_18,sc_19]

In [10]:
#create master df for street cars to concat all years onto one
sc_delay = pd.DataFrame()

for year in a:
    sc_delay = pd.concat([sc_delay, year])

In [11]:
#reset index inplace
sc_delay.reset_index(inplace = True, drop = True)

In [12]:
#peek into a df
sc_delay

Report Date  Route      Time      Day                 Location  \
0      2017-01-01    509  00:49:00   Sunday  Spadina and Queens Quay   
1      2017-01-01    501  01:07:00   Sunday    Lakeshore and Kipling   
2      2017-01-01    505  01:09:00   Sunday   Dundas and Howard Park   
3      2017-01-01    506  01:15:00   Sunday                High Park   
4      2017-01-01    506  01:18:00   Sunday               Parliament   
...           ...    ...       ...      ...                      ...   
41251  2019-12-31    505  20:00:00  Tuesday               Parliament   
41252  2019-12-31    505  20:32:00  Tuesday        Broadview Station   
41253  2019-12-31    501  22:32:00  Tuesday              Humber Loop   
41254  2019-12-31    506  23:20:00  Tuesday   Carlton and Parliament   
41255  2019-12-31    511  23:36:00  Tuesday         Bathurst Station   

                 Incident  Min Delay  Min Gap Direction  Vehicle  Incident ID  
0              Mechanical        7.0     14.0       N/B   4426.0          NaN  
1              Mechanical       32.0     44.0       W/B   4071.0          NaN  
2              Mechanical       10.0     20.0       W/B   4179.0          NaN  
3              Mechanical       10.0     20.0       E/B   4059.0          NaN  
4      Emergency Services       10.0     20.0       E/B   4106.0          NaN  
...                   ...        ...      ...       ...      ...          ...  
41251       General Delay       18.0     23.0       W/B   1262.0          NaN  
41252  Emergency Services        5.0     10.0       W/B   3104.0          NaN  
41253             Held By        5.0     10.0       E/B   4576.0          NaN  
41254       General Delay       20.0     26.0       E/B   4498.0          NaN  
41255             Held By        7.0     14.0       S/B   4583.0          NaN  

[41256 rows x 11 columns]

#### Master dataframe `sc_delay` is ready for data wrangling and cleaning

---

#### Standardize the data


In [13]:
#standardize Column Names first
sc_delay.columns = [i.replace(' ', '_').lower() for i in sc_delay.columns] 

In [14]:
[print(i,sc_delay[i].value_counts().shape) for i in sc_delay.columns]

report_date (1095,)
route (93,)
time (1442,)
day (7,)
location (9964,)
incident (9,)
min_delay (314,)
min_gap (316,)
direction (56,)
vehicle (2259,)
incident_id (9,)


[None, None, None, None, None, None, None, None, None, None, None]

In [15]:
#checking for null values
sc_delay.isna().sum()

report_date        0
route              0
time               0
day                0
location         152
incident           0
min_delay         69
min_gap          105
direction        251
vehicle         1908
incident_id    40367
dtype: int64

In [16]:
#dropping incident ID for a vast missing values issue
sc_delay = sc_delay.drop(['incident_id'], axis = 1)

In [17]:
#Not available vehicle number assigned to 9999
sc_delay['vehicle'].fillna(9999, inplace = True)

In [18]:
#Changing data type to integer for Vehicle coumn
sc_delay['vehicle'] = sc_delay['vehicle'].astype('int')

In [19]:
#null values for direction of vehicle assigned to unknown
sc_delay['direction'].fillna('Unknown', inplace = True)

In [20]:
#Not available location name is set as not available assigned to Not Available
sc_delay['location'].fillna('Not Available', inplace = True)

In [21]:
#Check remaining nulls
sc_delay.isna().sum()

report_date      0
route            0
time             0
day              0
location         0
incident         0
min_delay       69
min_gap        105
direction        0
vehicle          0
dtype: int64

In [22]:
#dropping all rows with null values now i.e delay or gap data only since that is the primary information
sc_delay = sc_delay.dropna()

In [23]:
sc_delay.reset_index(inplace= True, drop = True)

In [24]:
#Checking nulls and shape
print(sc_delay.isna().sum())
print('data shape: ', sc_delay.shape)

report_date    0
route          0
time           0
day            0
location       0
incident       0
min_delay      0
min_gap        0
direction      0
vehicle        0
dtype: int64
data shape:  (41139, 10)


In [25]:
sc_delay.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41139 entries, 0 to 41138
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   report_date  41139 non-null  datetime64[ns]
 1   route        41139 non-null  int64         
 2   time         41139 non-null  object        
 3   day          41139 non-null  object        
 4   location     41139 non-null  object        
 5   incident     41139 non-null  object        
 6   min_delay    41139 non-null  float64       
 7   min_gap      41139 non-null  float64       
 8   direction    41139 non-null  object        
 9   vehicle      41139 non-null  int32         
dtypes: datetime64[ns](1), float64(2), int32(1), int64(1), object(5)
memory usage: 3.0+ MB


## Addressing Route Column

In [26]:
#Check route info. There are only 14 operational routes for TTC
sc_delay['route'].value_counts().head(30)

501    10137
504     7694
506     5931
505     4883
512     3007
510     2371
511     2094
514     1094
509     1058
502      696
301      563
503      531
705      261
306      226
304      194
805      114
310       56
1         21
508       17
317       15
50        14
535       12
500        9
11         7
807        7
51         7
4          6
5          5
715        5
56         4
Name: route, dtype: int64

In [27]:
#Fixing potential typo errors

sc_delay['route'] = sc_delay['route'].replace([805, 705, 535, 5], 505)
sc_delay['route'] = sc_delay['route'].replace([1,51], 501)
sc_delay['route'] = sc_delay['route'].replace(11, 511)
sc_delay['route'] = sc_delay['route'].replace([54, 4, 54], 504)

#Dropping street car routes entry that is not a valid route number provided at:  https://www.ttc.ca/Routes/Streetcars.jsp 
sc_delay.drop(sc_delay[(sc_delay.route > 514) | ((sc_delay.route < 301) ) ].index, inplace=True)
sc_delay.drop(sc_delay[(sc_delay.route < 501) & ((sc_delay.route > 310) ) ].index, inplace=True)
sc_delay.drop(sc_delay[sc_delay.route == 307].index, inplace = True)

In [28]:
sc_delay.shape

(40988, 10)

In [29]:
sc_delay['route'].value_counts()

501    10165
504     7702
506     5931
505     5275
512     3007
510     2371
511     2101
514     1094
509     1058
502      696
301      563
503      531
306      226
304      194
310       56
508       17
513        1
Name: route, dtype: int64

### Cleaning `direction` column

In [30]:
# Standardize the Columns data
sc_delay['location'] = [i.replace(' ', '_').lower() for i in sc_delay['location']]
sc_delay['direction'] = [i.replace(' ', '_').lower() for i in sc_delay['direction']]

In [31]:
#checking 
sc_delay['direction'].value_counts()

w/b                    16588
e/b                    15776
n/b                     2775
s/b                     2750
b/w                     2495
unknown                  247
eb                       122
wb                       118
sb                        36
nb                        22
bw                        12
ew                        12
w                          3
e/w                        3
8                          2
5                          2
s                          1
4                          1
.                          1
2                          1
3                          1
15                         1
nbn                        1
relief_operator._ga        1
33                         1
n/s                        1
31087                      1
eb/wb                      1
-                          1
30                         1
wn                         1
see_also                   1
ee                         1
e5                         1
26            

In [32]:
#using mapping function to map all the signifcant unique values to their respective directions
mapping = {'e': 'east', 's':'south', 'w':'west', 'n':'north', 'e/b': 'east', 's/b':'south', 'w/b':'west', 'n/b':'north',
          'eb': 'east', 'sb':'south', 'wb':'west', 'nb':'north', 'be': 'east', 'bs':'south', 'bw':'west', 'bn':'north',
          'b/w':'west', 'ew': 'west'}
sc_delay['direction'] = sc_delay['direction'].map(mapping)

In [33]:
sc_delay['direction'].value_counts()

west     19228
east     15899
north     2797
south     2787
Name: direction, dtype: int64

In [34]:
sc_delay.shape

(40988, 10)

In [35]:
#242 unkown + 48 nulls in the direction column
sc_delay.isna().sum()

report_date      0
route            0
time             0
day              0
location         0
incident         0
min_delay        0
min_gap          0
direction      277
vehicle          0
dtype: int64

In [36]:
#null values for direction of vehicle assigned to unknown
sc_delay['direction'].fillna('Unknown', inplace = True)

### Cleaning `location` (Starting with 8420 Unique entries)

In [37]:
#Getting original value count
sc_delay['location'].value_counts()

leslie_barns           953
russell_yard           939
broadview_station      616
dundas_west_station    604
queen_and_connaught    506
                      ... 
duncan_/_richmond        1
queen_and_yongw          1
queena_and_bay           1
quen_at_mccaul           1
howardpark/parkside      1
Name: location, Length: 8420, dtype: int64

In [38]:
#Standardizing data in the location column
sc_delay['location'] = [i.replace('stn', 'station').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('.', '').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('stn', 'station').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('.', '').lower() for i in sc_delay['location']]

sc_delay['location'] = [i.replace(' ', '_').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('__', '_').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('___', '_').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('_-_', '_').lower() for i in sc_delay['location']]

#Fixing common typos
sc_delay['location'] = [i.replace('_nd_', 'and').lower() for i in sc_delay['location']]

# standardizing more common trends noticed
sc_delay['location'] = [i.replace('_and_' , '/').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('_/_', '/').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('_&_', '/').lower() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('_at_', '/').lower() for i in sc_delay['location']]

In [39]:
sc_delay['location'].value_counts().shape

(6183,)

We are down to 6183 unique values a decrease of almost 2300 values.

---

In [40]:
#Here we checked against different key words to find matching values to narrow down our results. 
# words used e.g. like = station or  like =  loop/queen/broadview/spadina/king/cne/dundas/clai/ronc/les/rus/nev
def checkexpression(expression):
    '''
    Takes a string and searches for similar entries in the locations dataframe created inside the functio.
    expression must be put inside ''. e.g checkexpression('expression')
    '''
    assert isinstance(expression, str), "Please add '' to your expresssion"
    locations = pd.DataFrame(sc_delay['location'].value_counts())
    #locations.reset_index(inplace = True)
    locations = locations.rename(columns={"index": "location", "location": "counts"})
    loop_list = locations.filter(like= expression , axis=0)
    return loop_list.head(30)

In [41]:
#redefine function everytime before running this to get updated results
checkexpression('dundas').head(20)

counts
dundas_west_station     845
broadview/dundas        214
dundas/bathurst         134
bathurst/dundas         118
dundas_w_station         95
dundas/broadview         95
dundas/spadina           81
spadina/dundas           77
dundas/yonge             75
dundas/dufferin          65
dundas/university        64
dundas/bay               64
dundas/lansdowne         63
dundas/parliament        62
dundas/ossington         60
dundas/howard_park       58
dundas                   56
roncesvalles/dundas      50
dundas/church            48
dundas/jarvis            45

#### Looking closely at the location column we can see most of the street car stops are listed as `street1/street2` since they are in no specific order we will organize them alphabetically to avoid any duplicates.

In [42]:
#Changing datatype to string
sc_delay['location'] = sc_delay['location'].astype('str')

In [43]:
#resetting index
sc_delay.reset_index(inplace= True, drop = True)

In [44]:
#separate them by '/' and then sort them
sc_delay.location = sc_delay.location.str.split('/')
for i in np.arange(0, (sc_delay.shape[0])):
    sc_delay.location[i].sort()

In [45]:
#join the sorted street names back 
c = sc_delay.location
sc_delay.location = [['/'.join(i)] for i in c]

In [46]:
#Convert array into a string
for i in np.arange(0, (sc_delay.shape[0])):
    sc_delay.location[i] = sc_delay.location[i][0]

C:\Users\safiu\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [47]:
sc_delay['location'].value_counts()

broadview_station                 1000
russell_yard                       968
leslie_barns                       955
queen/roncesvalles                 881
connaught/queen                    880
                                  ... 
wellington_westbound_of_church       1
lakeshore_blvd_w/marine_parade       1
galley_ave/roncesvalles_ave          1
lower_simcoe/qquay                   1
college_st/university_ave            1
Name: location, Length: 5567, dtype: int64

In [48]:
#Merging loop entries
sc_delay['location'] = sc_delay['location'].replace(['exibition_loop', 'exhibition_loop_to_bathurst_station', 'ex_loop', 'cne',
                                                    'exhibition','exhibtion_loop', 'cne_east_loop'],'exhibition_loop')
sc_delay['location'] = sc_delay['location'].replace(['distillary_loop', 'disterilly_loop', 'distrillery_loop'], 'distillery_loop')
sc_delay['location'] = sc_delay['location'].replace(['humberloop', 'humber_loop_to_neville_loop'], 'humber_loop')
sc_delay['location'] = sc_delay['location'].replace(['kingston_rd_loop', 'kingston_loop'], 'kingston_road_loop')
sc_delay['location'] = sc_delay['location'].replace(['gunn\'s_loop', 'guns_loop', 'gunns_loop_to_st_clair_station'],
                                                    'gunns_loop')
sc_delay['location'] = sc_delay['location'].replace(['lansdown_loop'], 'lansdowne_loop')
sc_delay['location'] = sc_delay['location'].replace(['queen_and_neville_loop', 'neville_park_loop_to_sunnyside_loop', 'neville_loop',
                                                     'neville_park','queen/neville_loop', 'neville'], 'neville_park_loop')
                                                    
sc_delay['location'] = sc_delay['location'].replace(['hpark_loop'], 'high_park_loop')
sc_delay['location'] = sc_delay['location'].replace(['charolette_loop'], 'charlotte_loop')
sc_delay['location'] = sc_delay['location'].replace(['birgham_loop'], 'bingham_loop')
sc_delay['location'] = sc_delay['location'].replace(['spadina_loop', 'spadina'], 'spadina_station')
sc_delay['location'] = sc_delay['location'].replace(['queens_quay_loop', 'union_tunnel', 'union_platform', 'union'], 'union_station')
sc_delay['location'] = sc_delay['location'].replace(['russel_yard', 'russell__yard'], 'russell_yard')
sc_delay['location'] = sc_delay['location'].replace(['leslie_barns_yard', 'leslie_yard'], 'leslie_barns')

In [49]:
sc_delay['location'].value_counts().shape

(5533,)

In [50]:
[print(i,sc_delay[i].value_counts().shape) for i in sc_delay.columns]

report_date (1095,)
route (17,)
time (1442,)
day (7,)
location (5533,)
incident (9,)
min_delay (313,)
min_gap (316,)
direction (5,)
vehicle (2250,)


[None, None, None, None, None, None, None, None, None, None]

In [51]:
sc_delay['location'] = sc_delay['location'].replace(['queen_and_broadview', 'broadview/queen', 'queen/broadview'], 'broadview_and_queen')
sc_delay['location'] = sc_delay['location'].replace(['spadina_and_king', 'king/spadina', 'spadina/king', 'spadina_at_king',
                                                    'king_&_spadina', 'king_/_spadina', 'king/spadina', 'king_at_spadina', 
                                                    'spadina_/_king', 'spadina_&_king'], 'king/spadina')
sc_delay['location'] = sc_delay['location'].replace('bathurst_and_king', 'king_and_bathurst')
sc_delay['location'] = sc_delay['location'].replace('gerrard_and_broadview', 'broadview/gerrard')
#Yard location for roncesvalles yard is same as roncesvalles and queen
sc_delay['location'] = sc_delay['location'].replace(['roncesvalles/queen', 'roncy_and_queen', 'roncesvalles_and_queen',
                                                     'queen_at_roncesvalles', 'queen_and_roncesvalles', 'roncy_yard', 'roncy',
                                                     'queen_and_roncy', 'queen/roncesvalles', 'roncesvalles', 'roncy', 
                                                     'roncy_&_queen', 'roncesvalles_&_queen', 'roncevalles'], 'roncesvalles_yard')
                                                   
sc_delay['location'] = sc_delay['location'].replace('longbranch_loop', 'long_branch_loop')
sc_delay['location'] = sc_delay['location'].replace('highpark_loop', 'high_park_loop')
sc_delay['location'] = sc_delay['location'].replace('coxwell_and_gerrard', 'coxwell/gerrard')

sc_delay['location'] = sc_delay['location'].replace('spadina_and_queen', 'queen/spadina')
sc_delay['location'] = sc_delay['location'].replace(['connaught_and_queen', 'queen_at_connaught', 'queen/connaught', 'queen_&_connaught'],
                                                    'connaught/queen')
sc_delay['location'] = sc_delay['location'].replace('howard_park_and_roncesvalles', 'howard_park/roncesvalles')
sc_delay['location'] = sc_delay['location'].replace(['st._clair_west_station', 'scw_station'], 'st_clair_west_station')
sc_delay['location'] = sc_delay['location'].replace('neville_loop', 'neville_park_loop')
sc_delay['location'] = sc_delay['location'].replace('college_and_spadina', 'college/spadina')
sc_delay['location'] = sc_delay['location'].replace('cne_loop', 'exhibition_loop')
sc_delay['location'] = sc_delay['location'].replace('queen_and_bathurst', 'bathurst/queen')
sc_delay['location'] = sc_delay['location'].replace(['dundas_west_to_broadview_station', 
                                                     'dundas_w_station_to_broadview_station',
                                                    'dundas_w_station-broadview_station', 'dundas_w_station-brdv_station',
                                                     'dundas_west_station_loop_to_broadview_station_loop'],
                                                    'dundas_west_station_to_broadview_station')
sc_delay['location'] = sc_delay['location'].replace(['broadview_station_loop_to_dundas_west_station_loop', 'broadview_station_to_dundas_w_station',
                                                     'broadview_to_dundas_west_station'],
                                                    'broadview_station_to_dundas_west_station')
sc_delay['location'] = sc_delay['location'].replace(['union_station_tunnel'], 'union_station')
sc_delay['location'] = sc_delay['location'].replace(['high_park_loop_to_main_station'], 'high_park_to_main_station')
sc_delay['location'] = sc_delay['location'].replace(['bathurst_station_to_cne_loop'], 'bathurst_station_to_cne')
sc_delay['location'] = sc_delay['location'].replace(['dundas/broadview', 'broadview_dundas'], 'broadview/dundas')
sc_delay['location'] = sc_delay['location'].replace(['dundas/bathurst'], 'bathurst/dundas')
sc_delay['location'] = sc_delay['location'].replace(['dundas/yonge','yonge/dundas'], 'dundas_station')



In [52]:
#null values for direction of vehicle assigned to unknown
sc_delay['direction'].fillna('Unknown', inplace = True)

In [53]:
sc_delay['location'].value_counts().shape

(5512,)

In [54]:
sc_delay

report_date  route      time      day             location  \
0      2017-01-01    509  00:49:00   Sunday  queens_quay/spadina   
1      2017-01-01    501  01:07:00   Sunday    kipling/lakeshore   
2      2017-01-01    505  01:09:00   Sunday   dundas/howard_park   
3      2017-01-01    506  01:15:00   Sunday            high_park   
4      2017-01-01    506  01:18:00   Sunday           parliament   
...           ...    ...       ...      ...                  ...   
40983  2019-12-31    505  20:00:00  Tuesday           parliament   
40984  2019-12-31    505  20:32:00  Tuesday    broadview_station   
40985  2019-12-31    501  22:32:00  Tuesday          humber_loop   
40986  2019-12-31    506  23:20:00  Tuesday   carlton/parliament   
40987  2019-12-31    511  23:36:00  Tuesday     bathurst_station   

                 incident  min_delay  min_gap direction  vehicle  
0              Mechanical        7.0     14.0     north     4426  
1              Mechanical       32.0     44.0      west     4071  
2              Mechanical       10.0     20.0      west     4179  
3              Mechanical       10.0     20.0      east     4059  
4      Emergency Services       10.0     20.0      east     4106  
...                   ...        ...      ...       ...      ...  
40983       General Delay       18.0     23.0      west     1262  
40984  Emergency Services        5.0     10.0      west     3104  
40985             Held By        5.0     10.0      east     4576  
40986       General Delay       20.0     26.0      east     4498  
40987             Held By        7.0     14.0     south     4583  

[40988 rows x 10 columns]

In [55]:
sc_delay['time'].value_counts().tail(10)

03:13:00               2
04:22:00               2
03:56:00               1
03:24:00               1
03:37:00               1
2017-05-27 00:00:00    1
2019-08-04 00:00:00    1
02:48:00               1
04:09:00               1
2019-10-14 00:00:00    1
Name: time, dtype: int64

In [56]:
sc_delay['time'] = sc_delay['time'].replace(['2020-05-23 00:00:00', '2017-05-27 00:00:00', '2019-08-04 00:00:00',
                                            '2020-05-21 00:00:00', '2019-10-14 00:00:00'], '00:00:00')

In [59]:
sc_delay

report_date  route      time      day             location  \
0      2017-01-01    509  00:49:00   Sunday  queens_quay/spadina   
1      2017-01-01    501  01:07:00   Sunday    kipling/lakeshore   
2      2017-01-01    505  01:09:00   Sunday   dundas/howard_park   
3      2017-01-01    506  01:15:00   Sunday            high_park   
4      2017-01-01    506  01:18:00   Sunday           parliament   
...           ...    ...       ...      ...                  ...   
40983  2019-12-31    505  20:00:00  Tuesday           parliament   
40984  2019-12-31    505  20:32:00  Tuesday    broadview_station   
40985  2019-12-31    501  22:32:00  Tuesday          humber_loop   
40986  2019-12-31    506  23:20:00  Tuesday   carlton/parliament   
40987  2019-12-31    511  23:36:00  Tuesday     bathurst_station   

                 incident  min_delay  min_gap direction  vehicle  
0              Mechanical        7.0     14.0     north     4426  
1              Mechanical       32.0     44.0      west     4071  
2              Mechanical       10.0     20.0      west     4179  
3              Mechanical       10.0     20.0      east     4059  
4      Emergency Services       10.0     20.0      east     4106  
...                   ...        ...      ...       ...      ...  
40983       General Delay       18.0     23.0      west     1262  
40984  Emergency Services        5.0     10.0      west     3104  
40985             Held By        5.0     10.0      east     4576  
40986       General Delay       20.0     26.0      east     4498  
40987             Held By        7.0     14.0     south     4583  

[40988 rows x 10 columns]

In [60]:
sc_delay['incident'].value_counts()

Mechanical               19008
Investigation             6194
Held By                   4568
Late Leaving Garage       3752
General Delay             3283
Emergency Services        1745
Utilized Off Route        1615
Diversion                  624
Overhead - Pantograph      199
Name: incident, dtype: int64

In [61]:
[print(i,sc_delay[i].value_counts().shape) for i in sc_delay.columns]

report_date (1095,)
route (17,)
time (1442,)
day (7,)
location (5512,)
incident (9,)
min_delay (313,)
min_gap (316,)
direction (5,)
vehicle (2250,)


[None, None, None, None, None, None, None, None, None, None]

In [62]:
# lets reorganize the columns and drop time column in the process to avoid repeition. 

sc_delay = sc_delay[['report_date','time','day','route','location','direction','min_delay','min_gap','incident','vehicle']]

In [63]:
#changing data types to integers
sc_delay['min_gap'] = sc_delay['min_gap'].astype('int')
sc_delay['min_delay'] = sc_delay['min_delay'].astype('int')

In [64]:
sc_delay.reset_index(inplace=True, drop = True)

In [65]:
sc_delay.head()

report_date      time     day  route             location direction  \
0  2017-01-01  00:49:00  Sunday    509  queens_quay/spadina     north   
1  2017-01-01  01:07:00  Sunday    501    kipling/lakeshore      west   
2  2017-01-01  01:09:00  Sunday    505   dundas/howard_park      west   
3  2017-01-01  01:15:00  Sunday    506            high_park      east   
4  2017-01-01  01:18:00  Sunday    506           parliament      east   

   min_delay  min_gap            incident  vehicle  
0          7       14          Mechanical     4426  
1         32       44          Mechanical     4071  
2         10       20          Mechanical     4179  
3         10       20          Mechanical     4059  
4         10       20  Emergency Services     4106

In [66]:
#Making Columns readable before writing it onto a csv
sc_delay['location'] = [i.replace('_', ' ').title() for i in sc_delay['location']]
sc_delay['location'] = [i.replace('/', ' / ').title() for i in sc_delay['location']]
sc_delay['direction'] = [i.title() for i in sc_delay['direction']]

In [67]:
#Last Check
sc_delay.head()

report_date      time     day  route               location direction  \
0  2017-01-01  00:49:00  Sunday    509  Queens Quay / Spadina     North   
1  2017-01-01  01:07:00  Sunday    501    Kipling / Lakeshore      West   
2  2017-01-01  01:09:00  Sunday    505   Dundas / Howard Park      West   
3  2017-01-01  01:15:00  Sunday    506              High Park      East   
4  2017-01-01  01:18:00  Sunday    506             Parliament      East   

   min_delay  min_gap            incident  vehicle  
0          7       14          Mechanical     4426  
1         32       44          Mechanical     4071  
2         10       20          Mechanical     4179  
3         10       20          Mechanical     4059  
4         10       20  Emergency Services     4106

## CLEAN DATA IS READY!

In [68]:
sc_delay.to_csv('sc_delay_17-20.csv')